File: api/src/APIBundle/Entity/Subaudience.php
<?php

namespace APIBundle\Entity;

use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\ORM\Mapping as ORM;

/**
 * @ORM\Entity()
 * @ORM\Table(name="subaudiences")
 */
class Subaudience
{
    /**
     * @ORM\Id
     * @ORM\Column(type="string")
     * @ORM\GeneratedValue(strategy="UUID")
     */
    protected $subaudience_id;

    /**
     * @ORM\Column(type="string")
     */
    protected $subaudience_name;

    /**
     * @ORM\ManyToMany(targetEntity="User", inversedBy="user_subaudiences")
     * @ORM\JoinTable(name="users_subaudiences",
     *      joinColumns={@ORM\JoinColumn(name="subaudience_id", referencedColumnName="subaudience_id", onDelete="CASCADE")},
     *      inverseJoinColumns={@ORM\JoinColumn(name="user_id", referencedColumnName="user_id", onDelete="CASCADE")}
     *      )
     * @var User[]
     */
    protected $subaudience_users;

    /**
     * @ORM\ManyToOne(targetEntity="Audience", inversedBy="audience_subaudiences")
     * @ORM\JoinColumn(name="subaudience_audience", referencedColumnName="audience_id", onDelete="CASCADE")
     * @var Audience
     */
    protected $subaudience_audience;

    /**
     * @ORM\Column(type="boolean")
     */
    protected $subaudience_enabled = true;

    /**
     * @ORM\ManyToMany(targetEntity="Inject", mappedBy="inject_subaudiences")
     * @var Inject[]
     */
    protected $subaudience_injects;

    protected $subaudience_exercise;

    public function __construct()
    {
        $this->subaudience_users = new ArrayCollection();
        $this->subaudience_injects = new ArrayCollection();
    }

    public function getSubaudienceId()
    {
        return $this->subaudience_id;
    }

    public function setSubaudienceId($id)
    {
        $this->subaudience_id = $id;
        return $this;
    }

    public function getSubaudienceName()
    {
        return $this->subaudience_name;
    }

    public function setSubaudienceName($name)
    {
        $this->subaudience_name = $name;
        return $this;
    }

    public function getSubaudienceUsers()
    {
        return $this->subaudience_users;
    }

    public function setSubaudienceUsers($users)
    {
        $this->subaudience_users = $users;
        return $this;
    }

    public function getSubaudienceAudience()
    {
        return $this->subaudience_audience;
    }

    public function setSubaudienceAudience($audience)
    {
        $this->subaudience_audience = $audience;
        return $this;
    }

    public function getSubaudienceEnabled()
    {
        return $this->subaudience_enabled;
    }

    public function setSubaudienceEnabled($enabled)
    {
        $this->subaudience_enabled = $enabled;
        return $this;
    }

    public function getSubaudienceInjects()
    {
        return $this->subaudience_injects;
    }

    public function setSubaudienceInjects($injects)
    {
        $this->subaudience_injects = $injects;
        return $this;
    }

    public function getSubaudienceExercise()
    {
        return $this->subaudience_exercise;
    }

    public function setSubaudienceExercise($exercise)
    {
        $this->subaudience_exercise = $exercise;
        return $this;
    }
}

File: openex-api/src/main/java/io/openex/database/model/SubAudience.java
package io.openex.database.model;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import io.openex.helper.MonoModelDeserializer;
import io.openex.helper.MultiModelDeserializer;
import org.hibernate.annotations.GenericGenerator;

import javax.persistence.*;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "subaudiences")
public class SubAudience implements Base {
    @Id
    @Column(name = "subaudience_id")
    @GeneratedValue(generator = "UUID")
    @GenericGenerator(name = "UUID", strategy = "org.hibernate.id.UUIDGenerator")
    @JsonProperty("subaudience_id")
    private String id;

    @Column(name = "subaudience_name")
    @JsonProperty("subaudience_name")
    private String name;

    @Column(name = "subaudience_enabled")
    @JsonProperty("subaudience_enabled")
    private boolean enabled;

    @ManyToOne
    @JoinColumn(name = "subaudience_audience")
    @JsonSerialize(using = MonoModelDeserializer.class)
    @JsonProperty("subaudience_audience")
    private Audience audience;

    @ManyToMany(fetch = FetchType.EAGER)
    @JoinTable(name = "users_subaudiences",
            joinColumns = @JoinColumn(name = "subaudience_id"),
            inverseJoinColumns = @JoinColumn(name = "user_id"))
    @JsonSerialize(using = MultiModelDeserializer.class)
    @JsonProperty("subaudience_users")
    private List<User> users = new ArrayList<>();

    // region transient
    @JsonProperty("subaudience_exercise")
    public String getExercise() {
        return getAudience().getExercise().getId();
    }

    // @JsonProperty("subaudience_injects")
    // public List<Inject<?>> getInjects() {
    //     return getAudience().getExercise().getEvents().stream()
    //             .flatMap(event -> event.getIncidents().stream())
    //             .flatMap(incident -> incident.getInjects().stream())
    //             .toList();
    // }
    // endregion

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public boolean isEnabled() {
        return enabled;
    }

    public void setEnabled(boolean enabled) {
        this.enabled = enabled;
    }

    public Audience getAudience() {
        return audience;
    }

    public void setAudience(Audience audience) {
        this.audience = audience;
    }

    public List<User> getUsers() {
        return users;
    }

    public void setUsers(List<User> users) {
        this.users = users;
    }
}
